File: src/main/java/com/intgrah/algorithms/sort/QuickSort.java
package com.intgrah.algorithms.sort;

import java.util.Comparator;
import java.util.Stack;

import static com.intgrah.algorithms.util.Util.swap;

public class QuickSort<T> extends Sorter<T> {

    public QuickSort(Comparator<T> ord) {
        super(ord);
    }

    @Override
    public void sort(T[] a) {
        Stack<Integer> los = new Stack<>();
        Stack<Integer> his = new Stack<>();
        los.push(0);
        his.push(a.length - 1);
        while (!los.empty()) {
            int lo = los.pop();
            int hi = his.pop();
            if (lo >= hi) {
                continue;
            }
            T p = a[(lo + hi) / 2];
            int lt = lo, eq = lo, gt = hi;
            while (eq <= gt) {
                int comp = ord.compare(a[eq], p);
                if (comp < 0) {
                    swap(a, eq++, lt++);
                } else if (comp > 0) {
                    swap(a, eq, gt--);
                } else {
                    eq++;
                }
            }
            los.push(lo);
            his.push(lt - 1);
            los.push(gt + 1);
            his.push(hi);
        }
    }

}

File: sort/quick_sort.py
from util import swap


def quick_sort(a: list) -> None:
    jobs = [(0, len(a) - 1)] # Stack of inclusive intervals
    while jobs:
        lo, hi = jobs.pop()
        if hi - lo <= 0:
            continue
        pivot = a[(lo + hi) // 2] # Middle element
        lt = eq = lo
        gt = hi
        while eq <= gt:
            if a[eq] < pivot:
                swap(a, eq, lt)
                lt += 1
                eq += 1
            elif a[eq] > pivot:
                swap(a, eq, gt)
                gt -= 1
            else:
                eq += 1
        jobs.append((lo, lt - 1))
        jobs.append((gt + 1, hi))
